accepted/00000-maintenance-windows.md

```diff
@@ -0,0 +1,409 @@
+
+
+### Assign a Maintenance Schedule to a System
+
+There could be multiple ways to assign systems to maintenance schedules.
+
+- In the systems details page we need to show the selected schedule with a view and a change option.
+- It should be possible to change the maintenance schedule for multiple systems using SSM
+- In the details of the Maintenance Schedule we could have a systems page where add/remove is possible.
+
+If a Maintenance Schedule is assigned and which one, should be visible in the systems details page.
+Implementation of XMLRPC functions for getting and setting the current maintenance schedule must be implemented.
+For setting a schedule should be possible for multiple systems using an array of system ids.
+
+## System in Maintenance Mode
+
+The current status of the system should be reflected in the User Interface and API.
+
+If a system is in Maintenance Mode, the User Interface could show a special icon [ || ] (pause symbol of an audio player).
+
+In the XMLRPC API a flag could be added to `system.getDetails()`.
+A special function `system.listSystemsInMaintenanceMode()` should be added as well.
+
+
+## Recurring Actions
+
+Recurring Actions are scheduled using a cron like syntax. It may not be aligned to a
+Maintenance Schedule defined via ICalendar.
+
+When it happens that an Recurring Action is executed outside of a maintenance window, it should be skipped.
+A warning should be logged and maybe a Notification raised.
+
+The customer need to take care to align the two schedules.
+
+Define recurring Maintenance Windows and align them with Recurring Actions.
+When a Maintenance Window in such a series was excluded and an alternative window added, a one time highstate action must be scheduled by the admin.
+
+Randomly appearing Maintenance Windows cannot be handled.
+
+
+### Traditional Clients
+
+We will not implement something special for traditional clients. For the first iteration we take care
+that actions are scheduled only inside of maintenance windows. This would apply also for traditional clients.
+
+As long as the traditional client uses osad, or `rhn_check` is executed inside of the maintenance window
+the feature would work for them as well.
+
+We will document in the manual, that for traditional clients actions are also executed outside of Maintenance Windows
+if `rhn_check` is executed only at a later point in time.
+
+## What 3rd party tools to create schedules
+
+We require customers to use 3rd party tools to create the ICalendar files.
+We need to test the most common used tools to generate them.
+
+- office 365 / outlook
+- google calendar
+- confluence ?
+
+For Uyuni users Linux tools might be interesting
+- korganizer
+- Lightning (Thunderbird Extension)
+- evolution
+
+
+## Notifications
+
+When a system is assigned to a maintenance schedule, every finished action should check, if the system is still in maintenance mode.
```

Comment:
Moreover, we need to check, if the system is still in the same maintenance **window**. What could happen, is that a long running action could overflow to a next maintenance window and everything would look ok.

Reply:
From my point of view not a hard requirement. At least not for the first iteration.
I think a typical customer define for a system only 1 window per week.
If that overflow to the next week without anybody notice it, they have a different problem.